src/schemaVisitor.ts

```diff
@@ -0,0 +1,464 @@
+    });
+
+    // Map from directive names to lists of SchemaDirectiveVisitor instances
+    // created while visiting the schema.
+    const createdVisitors: {
+      [directiveName: string]: SchemaDirectiveVisitor[]
+    } = Object.create(null);
+    Object.keys(directiveVisitors).forEach(directiveName => {
+      createdVisitors[directiveName] = [];
+    });
+
+    function visitorSelector(
+      type: VisitableSchemaType,
+      methodName: string,
+    ): SchemaDirectiveVisitor[] {
+      const visitors: SchemaDirectiveVisitor[] = [];
+      const directiveNodes = type.astNode && type.astNode.directives;
+      if (! directiveNodes) {
+        return visitors;
+      }
+
+      directiveNodes.forEach(directiveNode => {
+        const directiveName = directiveNode.name.value;
+        if (! hasOwn.call(directiveVisitors, directiveName)) {
+          return;
+        }
+
+        const visitorClass = directiveVisitors[directiveName];
+
+        // Avoid creating visitor objects if visitorClass does not override
+        // the visitor method named by methodName.
+        if (! visitorClass.implementsVisitorMethod(methodName)) {
+          return;
+        }
+
+        const decl = declaredDirectives[directiveName];
+        let args: { [key: string]: any };
+
+        if (decl) {
+          // If this directive was explicitly declared, use the declared
+          // argument types (and any default values) to check, coerce, and/or
+          // supply default values for the given arguments.
+          args = getArgumentValues(decl, directiveNode);
+        } else {
+          // If this directive was not explicitly declared, just convert the
+          // argument nodes to their corresponding JavaScript values.
+          args = Object.create(null);
+          directiveNode.arguments.forEach(arg => {
+            args[arg.name.value] = valueFromASTUntyped(arg.value);
+          });
+        }
+
+        // As foretold in comments near the top of the visitSchemaDirectives
+        // method, this is where instances of the SchemaDirectiveVisitor class
+        // get created and assigned names. While subclasses could override the
+        // constructor method, the constructor is marked as protected, so
+        // these are the only arguments that will ever be passed.
+        visitors.push(new visitorClass({
+          name: directiveName,
+          args,
+          visitedType: type,
+          schema,
+          context,
+        }));
+      });
+
+      if (visitors.length > 0) {
+        visitors.forEach(visitor => {
+          createdVisitors[visitor.name].push(visitor);
+        });
+      }
+
+      return visitors;
+    }
+
+    visitSchema(schema, visitorSelector);
+
+    return createdVisitors;
+  }
+
+  // Mark the constructor protected to enforce passing SchemaDirectiveVisitor
+  // subclasses (not instances) to visitSchemaDirectives.
+  protected constructor(config: {
+    name: string
+    args: { [name: string]: any }
```

Comment:
It would be nice for args and context to be type parameters on this so they can be validated in the implementing functions

Reply:
Ah, then they should probably be type parameters on the `SchemaDirectiveVisitor` class itself, right?